warehouse/accounts/views.py

```diff
@@ -735,6 +744,122 @@ def _get_two_factor_data(request, _redirect_to="/"):
+    except TokenMissing:
+        return _error(request._("Invalid token: no token supplied"))
+
+    # Check whether this token is being used correctly
+    if data.get("action") != "email-project-role-verify":
+        return _error(request._("Invalid token: not a collaboration invitation token"))
+
+    user = user_service.get_user(data.get("user_id"))
+    if user != request.user:
+        return _error(request._("Invite is not valid"))
+
+    project = (
+        request.db.query(Project).filter(Project.id == data.get("project_id")).one()
+    )
+    desired_role = data.get("desired_role")
+
+    role_invite = (
+        request.db.query(RoleInvitation)
+        .filter(RoleInvitation.project == project)
+        .filter(RoleInvitation.user == user)
+        .one()
+    )
+
+    if role_invite.invite_status == RoleInvitationStatus.Revoked.value:
+        return _error(request._("Role invitation has been revoked."))
+    elif role_invite.invite_status == RoleInvitationStatus.Accepted.value:
+        return _error(request._("Role invitation is already accepted."))
+
+    role_invite.invite_status = RoleInvitationStatus.Accepted.value
+    request.db.add(Role(user=user, project=project, role_name=desired_role))
+    request.db.add(
+        JournalEntry(
+            name=project.name,
+            action=f"accepted {desired_role} {user.username}",
+            submitted_by=request.user,
+            submitted_from=request.remote_addr,
+        )
+    )
+    project.record_event(
+        tag="project:role:accepted",
+        ip_address=request.remote_addr,
+        additional={
+            "submitted_by": request.user.username,
+            "role_name": desired_role,
+            "target_user": user.username,
+        },
+    )
+    user.record_event(
+        tag="account:role:accepted",
+        ip_address=request.remote_addr,
+        additional={
+            "submitted_by": request.user.username,
+            "project_name": project.name,
+            "role_name": desired_role,
+        },
+    )
+
+    owner_roles = (
+        request.db.query(Role)
+        .filter(Role.project == project)
+        .filter(Role.role_name == "Owner")
+        .all()
+    )
+    owner_users = {owner.user for owner in owner_roles}
+
+    # Don't send email to new user if they are now an owner
+    owner_users.discard(user)
+
+    submitter_user = user_service.get_user(data.get("submitter_id"))
+    send_collaborator_added_email(
+        request,
+        owner_users,
+        user=user,
+        submitter=submitter_user,
+        project_name=project.name,
+        role=desired_role,
+    )
+
+    send_added_as_collaborator_email(
+        request,
+        user,
+        submitter=submitter_user,
+        project_name=project.name,
+        role=desired_role,
+    )
+
+    request.session.flash(
+        request._(
+            "Added collaborator '${username}'", mapping={"username": user.username},
```

Comment:
This is the message displayed to the user after accepting an invitation, might be better for this to say something like:

> You are now Owner/Maintainer of the 'project_name' project